iceberg/src/main/scala/org/apache/spark/sql/delta/icebergShaded/IcebergTransactionUtils.scala

```diff
@@ -210,6 +255,109 @@ object IcebergTransactionUtils
     builder
   }
 
+  private[delta] def convertFileAction(
+      f: FileAction,
+      tablePath: Path,
+      partitionSpec: PartitionSpec,
+      logicalToPhysicalPartitionNames: Map[String, String],
+      schema: StructType): DataFiles.Builder = {
+    val absPath = canonicalizeFilePath(f, tablePath)
+
+    var builder = DataFiles
+      .builder(partitionSpec)
+      .withPath(absPath)
+      .withFileSizeInBytes(f.getFileSize)
+      .withFormat(FileFormat.PARQUET)
+    val nameToDataTypes = schema.fields.map(f => f.name -> f.dataType).toMap
+
+    if (partitionSpec.isPartitioned) {
+      val ICEBERG_NULL_PARTITION_VALUE = "__HIVE_DEFAULT_PARTITION__"
+      val partitionPath = partitionSpec.fields()
+      val partitionVals = new Array[Any](partitionSpec.fields().size())
+      for (i <- partitionVals.indices) {
+        val logicalPartCol = partitionPath.get(i).name()
+        val physicalPartKey = logicalToPhysicalPartitionNames(logicalPartCol)
+        // ICEBERG_NULL_PARTITION_VALUE is referred in Iceberg lib to mark NULL partition value
+        val partValue = Option(f.partitionValues(physicalPartKey))
+          .getOrElse(ICEBERG_NULL_PARTITION_VALUE)
+        val partitionColumnDataType = nameToDataTypes(logicalPartCol)
+        val deltaPartitionValue =
+          deserializeDeltaPartitionValue(partitionColumnDataType, partValue)
+        val icebergPartitionValue =
+          deltaPartitionValueToIcebergPartitionValue(deltaPartitionValue, partitionColumnDataType)
+        partitionVals(i) = icebergPartitionValue
+      }
```

Comment:
Also is it possible that we directly use the type information from `PartitionSpec` to do the conversion?

Reply:
>Why don't we directly convert string to icebergPartitionValue?
So this is essentially what was being done here before via the `withPartitionPath` API which takes in a string hive style path and does a conversion to an Iceberg partition value. The issue is that in Iceberg, partition values to string are considered 1 way (it's not well defined to go from string to Iceberg partition value from an iceberg perspective). There were some utilities exposed by Iceberg for the string conversion for functions like Hive table migration but again it's not a well defined conversion.

Instead the recommended approach is to construct explicit partition values via `StructLike`. The way we construct those explicit partition values is by converting partition string to a delta partition value which is defined by delta protocol. Then we can convert into explicit Structlike as per Iceberg's defined type API https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/types/Type.java#L32

Reply:
>Also is it possible that we directly use the type information from PartitionSpec to do the conversion?

Not from PartitionSpec itself, we'd need the Iceberg schema as well. PartitionSpec will have the field ID for the transform but it won't have the type information, we'd need to lookup the type from the schema.

Also I think since we're doing the string -> delta conversions since that is defined by the protocol we'll want to use the type information from the delta schema so that we can do that logic a bit more cleanly?

Reply:
Thanks for the reply. 
> The way we construct those explicit partition values is by converting partition string to a delta partition value which is defined by delta protocol. Then we can convert into explicit `Structlike` as per Iceberg's defined type API

> Also I think since we're doing the string -> delta conversions since that is defined by the protocol we'll want to use the type information from the delta schema so that we can do that logic a bit more cleanly?

I am totally fine with this two-step process as a design. However, as these are internal functions not being used anywhere else, I think we can merge the two functions together and having only one convert function such as `stringToIcebergPartitionValue(elemType, partitionVal, version)`. If you want to expose `deserializeDeltaPartitionValue` later as public API we can then decompose it as an independent function.

> There's no type info on the PartitionSpec itself, we'd need the Iceberg schema as well.

I think we have `PartitionSpec.schema`, which contains `Types.StructType struct`. This one should contain type information? Please correct me if this is not the case.

Reply:
>I think we have PartitionSpec.schema, which contains Types.StructType struct. This one should contain type information? Please correct me if this is not the case.

Ah correct, I forgot the schema that's bound to the PartitionSpec. That's an Iceberg Schema which will have the field ID to type information. But I think the below still applies:

>Also I think since we're doing the string -> delta conversions since that is defined by the protocol we'll want to use the type information from the delta schema so that we can do that logic a bit more cleanly?

If we use the Iceberg type information the logic for deserialization of the Delta partition values would be done by mapping Iceberg type information into Delta deserialization which is possible but imo to a reader seems more difficult to understand the logic compared to just using the delta types? If this is about minimizing the API changes, we still need to pass in the Delta snapshot details for getting details like versions in the error messages for the failure cases.

Reply:
>I am totally fine with this two-step process as a design. However, as these are internal functions not being used anywhere else, I think we can merge the two functions together and having only one convert function such as stringToIcebergPartitionValue(elemType, partitionVal, version). If you want to expose deserializeDeltaPartitionValue later as public API we can then decompose it as an independent function.

Sure that makes sense, will update!

Reply:
Cleaned it up, great suggestion @harperjiang it def is cleaner now!